frankenphp.c

```diff
@@ -663,7 +664,24 @@ static void frankenphp_register_variables(zval *track_vars_array) {
   /* In CGI mode, we consider the environment to be a part of the server
    * variables
    */
-  php_import_environment_variables(track_vars_array);
+
+  frankenphp_server_context *ctx = SG(server_context);
+
+  /* in non-worker mode we import the os environment regularly */
+  if (!ctx->has_main_request) {
+    php_import_environment_variables(track_vars_array);
+    go_register_variables(thread_index, track_vars_array);
+    return;
+  }
+
+  /* In worker mode we cache the os environment */
+  if (os_environment == NULL) {
+    os_environment = malloc(sizeof(zval));
+    array_init(os_environment);
+    php_import_environment_variables(os_environment);
+  }
+  zend_hash_copy(Z_ARR_P(track_vars_array), Z_ARR_P(os_environment),
+                 (copy_ctor_func_t)zval_add_ref);
```

Comment:
Any reason why this is on a separate line?

I'm not sure `zval_add_ref` is the right copy constructor to use here, I suspect it should be `NULL`. Have you tried running php in debug mode?

Reply:
`clang-format` will put it on a separate line if that's what you mean.

Interesting, I'll try setting up debug mode 👍

Reply:
I'm pretty sure it has to be `zval_add_ref` though, otherwise the zvals in the array will be garbage collected when superglobals are destroyed (with NULL I'm getting a segfault).

Reply:
Ah, yeah, I guess that makes sense. Running php in debug mode should point out if there is any php memory leak as well.

Reply:
Oh thanks for the info! The `dev.Dockerfile` is already running in debug mode I think. I was getting `memory leaks detected` warnings before.

What confuses me a bit is the difference between `zend_array_destroy` and `zval_ptr_dtor`, they both remove the memory leaks warnings, but I'm not sure if they're doing the same thing.

Reply:
IIRC, `zval_ptr_dtor` will call the registered destructor (aka, `zend_array_destroy`). The former is 'safer' in that you are operating on a zval that happens to point to an array.